### PR TITLE
CrateLocator refactorings

### PR DESCRIPTION
**Pull request by bjorn3**
This makes the `CrateLocator` a lot cleaner IMHO and much more self-contained. The last commit removes `extra_filename` from the crate metadata. This is an **insta-stable** change as it allows a crate like `libfoo-abc.rlib` to be used as dependency and then be renamed as `libfoo-bcd.rlib` while still being found as indirect dependency. This may reduce performance when there are a lot of versions of the same crate available as the extra filename won't be used to do an early rejection of crates before trying to load metadata, but it makes the logic to find the right filename a lot cleaner.

### CODE REVIEW

**Comment by bjorn3** [comment]
I'm guessing that this failure was caused by the extra filename changes.

**Comment by petrochenkov** [comment]
I'm going to review because "continue refactoring crate locator" is an item from my work queue on which I was going to work myself (cc https://github.com/rust-lang/rust/issues/56935#issuecomment-653935780).

**Comment by bjorn3** [comment]
I forgot an ! in the `only_needs_metadata` calculation.

**Comment by bjorn3** [comment]
I think I will revert "Don't pass Session to CrateError::report". Unlike with `CrateLocator` I don't think there are much benefits to CrateError::report not taking a Session after all.

**Comment by petrochenkov** [comment]
>This may reduce performance when there are a lot of versions of the same crate available as the extra filename won't be used to do an early rejection of crates before trying to load metadata, but it makes the logic to find the right filename a lot cleaner.

As far as I understand, the crates will now be (usually) rejected via hash kept in metadata instead of the extra filename piece (which is also some kind of hash if cargo is used).

Besides many versions of the same crate the perf regression can hit many crates with the same prefix like `librustc-abcd.rlib`/`librustc_foo-efgh.rlib`/`librustc_bar-ijkl.rlib`/etc, and the search for the library `rustc` will now need to open all the `librustc*` crate files and read hashes from them, because we no longer rely on knowing `-abcd`.

**Comment by bjorn3** [comment]
> As far as I understand, the crates will now be (usually) rejected via hash kept in metadata instead of the extra filename piece (which is also some kind of hash if cargo is used).

That and the file prefix depending in the crate name.

> and the search for the library rustc will now need to open all the librustc* crate files and read hashes from them, because we no longer rely on knowing -abcd.

Didn't think about that. Would have been nice if the extra filename was required to start with `-` to clearly separate the crate name and extra filenane. It is now a breaking change to change this unfortunately even though cargo always does this.

File: compiler/rustc_metadata/src/locator.rs [hunk 1]
```diff
@@ -242,6 +243,7 @@ crate struct CrateLocator<'a> {
     // Immutable per-session configuration.
     sess: &'a Session,
     metadata_loader: &'a dyn MetadataLoader,
+    current_crate: Option<&'a str>,
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/locator.rs:246]: `current_crate` isn't used by `CrateLocator`, only by `CrateError::report`, so I'd pass it to that function instead (together with the rest of the `Session` as you wrote in https://github.com/rust-lang/rust/pull/88538#issuecomment-909456333).

**Comment by bjorn3** [comment]
I think the regression is neglectable.

> and the search for the library rustc will now need to open all the librustc* crate files and read hashes from them, because we no longer rely on knowing -abcd.

I think a perf regression for this specific case can be prevented by checking with `-` as extra prefix first before checking without extra prefix. Assuming that the extra filename starts with a `-` most of the time this would prevent most of the unnecessary lookup. It would still do the unnecessary lookup if the extra filenames don't start with a `-`. Speaking of which I actually don't think this commit changes behavior. If the file is not found with the specified extra filename as extra prefix, it will already try again without extra prefix, which matches all crates with the right crate name independent of the extra filename.

File: compiler/rustc_metadata/src/locator.rs [hunk 2]
```diff
@@ -256,6 +255,7 @@ crate struct CrateLocator<'a> {
     pub triple: TargetTriple,
     pub filesearch: FileSearch<'a>,
     root: Option<&'a CratePaths>,
+    only_needs_metadata: bool,
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/locator.rs:255]: Looks like the previous placement was actually right, `only_needs_metadata` is only based on the session and not on the currently searched crate. All other fields in the "per-search configuration" apply to the specific crate that we are searching right now and can be different for other searches.

**Comment by petrochenkov** [comment]
>Speaking of which I actually don't think this commit changes behavior. If the file is not found with the specified extra filename as extra prefix, it will already try again without extra prefix, which matches all crates with the right crate name independent of the extra filename.

It can change behavior in theory if the file *is* found with the specified extra prefix.
With the old logic we stop searching, with the new logic we continue searching and can potentially report "multiple candidates" errors. This is not a very realistic case because for the error to be reported hashes must match, but the extra search can take some time.

**Comment by petrochenkov** [comment]
I'm still not sure about no longer keeping the `-C extra_filename` hints, they are cheap to keep and can potentially speed up crate search.
I need a second opinion, from cargo people maybe. cc @ehuss @alexcrichton

**Comment by bjorn3** [comment]
> This is not a very realistic case because for the error to be reported hashes must match, but the extra search can take some time.

Doesn't the search stop as soon a crate with the right SVH was found if a required SVH was specified?

**Comment by petrochenkov** [comment]
@bjorn3 
>Doesn't the search stop as soon a crate with the right SVH was found if a required SVH was specified?

It does not? I've just looked through the code, but haven't noticed any such early return logic.

File: compiler/rustc_metadata/src/locator.rs (hunk 1 above)
Comment (bjorn3) [reply compiler/rustc_metadata/src/locator.rs]: Reverted this commit.

File: compiler/rustc_metadata/src/locator.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_metadata/src/locator.rs]: Reverted this commit except the `PathBuf`->`Path` change.

**Comment by bjorn3** [comment]
I will squash the reverts after review.

**Comment by alexcrichton** [comment]
I don't know the specifics of what this PR is about, and I can't find a necessarily clear question that's being asked, but I can try to offer some thoughts anyway. From [the comment above](https://github.com/rust-lang/rust/pull/88538#issuecomment-909665421) it sounds like previously rustc is taking `-Cextra-filename` into account when searching for dependencies, first looking at rlibs with a matching `-Cextra-filename` and succeeding quickly if found. If that's the case and it's now being removed, I would caution against anything that would cause more candidate rlibs to be searched. Cargo dumps everything into `deps` for dependencies, and with rebuilds across features and Rust versions it's not uncommon for development directories to have hundreds of copies of `libfoo-xxxxx.rlib`. If all compilations have to search through all those rlibs to reject them that means that incremental compiles are going to be that much slower. This sort of test case is unlikely to really be exercised on perf.r-l.o since it's always working with pristine directories and it's relatively rare that multiple `libfoo-xxxx.rlib` files will appear in one crate graph.

**Comment by bjorn3** [comment]
I reverted the extra-filename commit. While I think it would be nice to drop extra-filename, the downsides don't weight up against it. At least without some kind of cache to map from svh to filename, but that introduces much more complexity elsewhere.

**Comment by petrochenkov** [comment]
r=me with commits appropriately squashed.

**Comment by bjorn3** [comment]
Squashed all reverts and the bugfix commit.

@bors r=petrochenkov

**Pull request merged**